crypto/thread/internal.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2019-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/configuration.h>
+#include <openssl/e_os2.h>
+#include <openssl/crypto.h>
+
+#if defined(OPENSSL_THREADS)
+
+# include <internal/thread.h>
+# include <internal/thread_arch.h>
+# include "openssl_threads.h"
+# include "worker.h"
+
+static int thread_start_now(OSSL_LIB_CTX *, CRYPTO_TASK *);
+static ossl_inline void thread_schedule(OSSL_LIB_CTX_THREADS *, CRYPTO_TASK *);
+static ossl_inline int thread_can_spawn(OSSL_LIB_CTX_THREADS *);
+
+void *crypto_thread_start(OSSL_LIB_CTX *ctx, CRYPTO_THREAD_ROUTINE start,
+                          void *data, int options)
+{
+    CRYPTO_TASK *task;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL || tdata->enabled == 0)
+        return NULL;
+
+    task = crypto_task_new(start, data);
+    if (task == NULL)
+        return NULL;
+
+    return crypto_thread_start_task(ctx, task, options);
+}
+
+void *crypto_thread_start_task(OSSL_LIB_CTX *ctx, void *vtask, int options)
+{
+    CRYPTO_TASK *task;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    task = (CRYPTO_TASK *)vtask;
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL || vtask == NULL)
+        return NULL;
+
+    CRYPTO_MUTEX_lock(task->lock);
+    {
+        CRYPTO_MUTEX_lock(tdata->lock);
+        {
+            task->id = tdata->tasks.created_tasks++;
+            if (thread_can_spawn(tdata) == 0)
+                thread_schedule(tdata, task);
+            else if (thread_start_now(ctx, task) == 0)
+                thread_schedule(tdata, task);
+        }
+        CRYPTO_MUTEX_unlock(tdata->lock);
+
+        while ((options & CRYPTO_THREAD_START_AWAIT) != 0) {
+            CRYPTO_mem_barrier();
```

Comment:
It would be useful to document in a comment why we need a mem barrier when we call this.

Reply:
I serialized crypto_thread_start_task, so this section got removed. Marking as resolved.